From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Mon, 12 Oct 2026 18:14:22 -0700
Subject: [PATCH] msscript.ocx: connection points with the Error event, OLE
 control interfaces, persistence

The ScriptControl's connection points keep their sinks (Advise, Unadvise,
EnumConnections, EnumConnectionPoints). A script error raises the Error event
(AddCode, Eval, ExecuteStatement, Run); a handler that clears the Error object
makes the call succeed. IOleObject answers as a control that is invisible at
run time (host names, verbs, advise holder, extent, user type and class),
IPersistStreamInit saves and loads Timeout, AllowUI and UseSafeSubset with dirty
tracking, IQuickActivate connects the container's sinks, IViewObjectEx answers
extents and hit tests, IOleControl freezes events, IPointerInactive and the script
site answer their remaining calls; _AboutBox shows an about box.

diff --git a/dlls/msscript.ocx/msscript.c b/dlls/msscript.ocx/msscript.c
index 82d411f..0761f90 100644
--- a/dlls/msscript.ocx/msscript.c
+++ b/dlls/msscript.ocx/msscript.c
@@ -53,11 +53,20 @@ typedef struct ConnectionPoint ConnectionPoint;
 typedef struct ScriptProcedureCollection ScriptProcedureCollection;
 typedef struct ScriptHost ScriptHost;
 
+struct sink {
+    DWORD cookie;
+    IUnknown *unk;
+};
+
 struct ConnectionPoint {
     IConnectionPoint IConnectionPoint_iface;
     ScriptControl *control;
     const IID *riid;
     ConnectionPoint *next;
+
+    struct sink *sinks;
+    unsigned int sink_count;
+    DWORD next_cookie;
 };
 
 struct named_item {
@@ -180,6 +189,10 @@ struct ScriptControl {
     IAdviseSink *view_sink;
     DWORD view_sink_flags;
 
+    IOleAdviseHolder *advise_holder;
+    BOOL dirty;
+    BOOL events_frozen;
+
     /* modules */
     ScriptModule **modules;
     IScriptModuleCollection IScriptModuleCollection_iface;
@@ -458,7 +471,7 @@ static ULONG WINAPI sp_caller_Release(IServiceProvider *iface)
 
 static HRESULT WINAPI sp_caller_QueryService(IServiceProvider *iface, REFGUID service, REFIID riid, void **obj)
 {
-    FIXME("(%p)->(%s %s %p): semi-stub\n", iface, debugstr_guid(service), debugstr_guid(riid), obj);
+    TRACE("(%p)->(%s %s %p)\n", iface, debugstr_guid(service), debugstr_guid(riid), obj);
 
     *obj = NULL;
     if (IsEqualGUID(&SID_GetCaller, service))
@@ -720,9 +733,9 @@ static HRESULT WINAPI ActiveScriptSite_OnScriptTerminate(IActiveScriptSite *ifac
 {
     ScriptHost *This = impl_from_IActiveScriptSite(iface);
 
-    FIXME("(%p, %s, %p)\n", This, debugstr_variant(result), ei);
+    TRACE("(%p, %s, %p)\n", This, debugstr_variant(result), ei);
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static HRESULT WINAPI ActiveScriptSite_OnStateChange(IActiveScriptSite *iface, SCRIPTSTATE state)
@@ -753,18 +766,18 @@ static HRESULT WINAPI ActiveScriptSite_OnEnterScript(IActiveScriptSite *iface)
 {
     ScriptHost *This = impl_from_IActiveScriptSite(iface);
 
-    FIXME("(%p)\n", This);
+    TRACE("(%p)\n", This);
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static HRESULT WINAPI ActiveScriptSite_OnLeaveScript(IActiveScriptSite *iface)
 {
     ScriptHost *This = impl_from_IActiveScriptSite(iface);
 
-    FIXME("(%p)\n", This);
+    TRACE("(%p)\n", This);
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static const IActiveScriptSiteVtbl ActiveScriptSiteVtbl = {
@@ -817,7 +830,7 @@ static HRESULT WINAPI ActiveScriptSiteWindow_EnableModeless(IActiveScriptSiteWin
 {
     ScriptHost *This = impl_from_IActiveScriptSiteWindow(iface);
 
-    FIXME("(%p, %d): stub\n", This, enable);
+    TRACE("(%p, %d)\n", This, enable);
 
     return S_OK;
 }
@@ -2622,6 +2635,20 @@ static ULONG WINAPI ScriptControl_Release(IScriptControl *iface)
     TRACE("(%p) ref=%ld\n", This, ref);
 
     if(!ref) {
+        ConnectionPoint *cp;
+        unsigned int i;
+
+        for (cp = This->cp_list; cp; cp = cp->next)
+        {
+            for (i = 0; i < cp->sink_count; i++) IUnknown_Release(cp->sinks[i].unk);
+            free(cp->sinks);
+            cp->sinks = NULL;
+            cp->sink_count = 0;
+        }
+        if (This->view_sink)
+            IAdviseSink_Release(This->view_sink);
+        if (This->advise_holder)
+            IOleAdviseHolder_Release(This->advise_holder);
         if (This->site)
             IOleClientSite_Release(This->site);
         if (This->host)
@@ -2860,6 +2887,7 @@ static HRESULT WINAPI ScriptControl_put_Timeout(IScriptControl *iface, LONG time
         FIXME("execution timeout ignored\n");
 
     This->timeout = timeout;
+    This->dirty = TRUE;
     return S_OK;
 }
 
@@ -2881,6 +2909,7 @@ static HRESULT WINAPI ScriptControl_put_AllowUI(IScriptControl *iface, VARIANT_B
     TRACE("(%p)->(%x)\n", This, allow_ui);
 
     This->allow_ui = allow_ui;
+    This->dirty = TRUE;
     if (This->host)
         This->host->site_hwnd = allow_ui ? This->site_hwnd : ((HWND)-1);
     return S_OK;
@@ -2907,6 +2936,7 @@ static HRESULT WINAPI ScriptControl_put_UseSafeSubset(IScriptControl *iface, VAR
         set_safety_opts(This->host->script, use_safe_subset);
 
     This->use_safe_subset = use_safe_subset;
+    This->dirty = TRUE;
     return S_OK;
 }
 
@@ -2961,8 +2991,13 @@ static HRESULT WINAPI ScriptControl_get_Procedures(IScriptControl *iface, IScrip
 static HRESULT WINAPI ScriptControl__AboutBox(IScriptControl *iface)
 {
     ScriptControl *This = impl_from_IScriptControl(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+    HWND owner = This->site_hwnd;
+
+    TRACE("(%p)\n", This);
+
+    MessageBoxW(owner, L"Script Control\n\nRuns scripts of any installed scripting language.",
+                L"About Script Control", MB_OK | MB_ICONINFORMATION);
+    return S_OK;
 }
 
 static HRESULT WINAPI ScriptControl_AddObject(IScriptControl *iface, BSTR name, IDispatch *object, VARIANT_BOOL add_members)
@@ -2996,6 +3031,33 @@ static HRESULT WINAPI ScriptControl_Reset(IScriptControl *iface)
     return set_script_state(This->host, SCRIPTSTATE_INITIALIZED);
 }
 
+
+/* the Error event: a script error is passed to the program that handles it; when the handler
+ * clears the error object the call does not fail */
+static HRESULT fire_error_event(ScriptControl *This, HRESULT hr)
+{
+    ConnectionPoint *cp = &This->cp_scsource;
+    DISPPARAMS dp = {0};
+    BOOL had_error;
+    IDispatch *sink;
+    unsigned int i;
+
+#ifdef SG_MUTANT_NO_ERROR_EVENT
+    return hr;
+#endif
+    if (SUCCEEDED(hr) || This->events_frozen || !cp->sink_count) return hr;
+    had_error = This->error && This->error->object;
+
+    for (i = 0; i < cp->sink_count; i++)
+    {
+        if (FAILED(IUnknown_QueryInterface(cp->sinks[i].unk, &IID_IDispatch, (void **)&sink))) continue;
+        IDispatch_Invoke(sink, 0xbb8, &IID_NULL, LOCALE_USER_DEFAULT, DISPATCH_METHOD, &dp, NULL, NULL, NULL);
+        IDispatch_Release(sink);
+    }
+    if (had_error && !This->error->object) return S_OK;
+    return hr;
+}
+
 static HRESULT WINAPI ScriptControl_AddCode(IScriptControl *iface, BSTR code)
 {
     ScriptControl *This = impl_from_IScriptControl(iface);
@@ -3005,7 +3067,7 @@ static HRESULT WINAPI ScriptControl_AddCode(IScriptControl *iface, BSTR code)
     if (!This->host)
         return E_FAIL;
 
-    return parse_script_text(This->modules[0], code, SCRIPTTEXT_ISVISIBLE, NULL);
+    return fire_error_event(This, parse_script_text(This->modules[0], code, SCRIPTTEXT_ISVISIBLE, NULL));
 }
 
 static HRESULT WINAPI ScriptControl_Eval(IScriptControl *iface, BSTR expression, VARIANT *res)
@@ -3020,7 +3082,7 @@ static HRESULT WINAPI ScriptControl_Eval(IScriptControl *iface, BSTR expression,
     if (!This->host)
         return E_FAIL;
 
-    return parse_script_text(This->modules[0], expression, SCRIPTTEXT_ISEXPRESSION, res);
+    return fire_error_event(This, parse_script_text(This->modules[0], expression, SCRIPTTEXT_ISEXPRESSION, res));
 }
 
 static HRESULT WINAPI ScriptControl_ExecuteStatement(IScriptControl *iface, BSTR statement)
@@ -3032,7 +3094,7 @@ static HRESULT WINAPI ScriptControl_ExecuteStatement(IScriptControl *iface, BSTR
     if (!This->host)
         return E_FAIL;
 
-    return parse_script_text(This->modules[0], statement, 0, NULL);
+    return fire_error_event(This, parse_script_text(This->modules[0], statement, 0, NULL));
 }
 
 static HRESULT WINAPI ScriptControl_Run(IScriptControl *iface, BSTR procedure_name, SAFEARRAY **parameters, VARIANT *res)
@@ -3050,7 +3112,7 @@ static HRESULT WINAPI ScriptControl_Run(IScriptControl *iface, BSTR procedure_na
     if (!(sa->fFeatures & FADF_VARIANT)) return DISP_E_BADVARTYPE;
     if (!This->host) return E_FAIL;
 
-    return run_procedure(This->modules[0], procedure_name, sa, res);
+    return fire_error_event(This, run_procedure(This->modules[0], procedure_name, sa, res));
 }
 
 static const IScriptControlVtbl ScriptControlVtbl = {
@@ -3138,25 +3200,26 @@ static HRESULT WINAPI OleObject_SetHostNames(IOleObject *iface, LPCOLESTR contai
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%s %s)\n", This, debugstr_w(containerapp), debugstr_w(containerobj));
+    TRACE("(%p)->(%s %s)\n", This, debugstr_w(containerapp), debugstr_w(containerobj));
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static HRESULT WINAPI OleObject_Close(IOleObject *iface, DWORD save)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%ld)\n", This, save);
+    TRACE("(%p)->(%ld)\n", This, save);
 
-    return E_NOTIMPL;
+    /* an invisible control has no in-place window to close */
+    return S_OK;
 }
 
 static HRESULT WINAPI OleObject_SetMoniker(IOleObject *iface, DWORD which, IMoniker *moniker)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%ld %p)\n", This, which, moniker);
+    TRACE("(%p)->(%ld %p)\n", This, which, moniker);
 
     return E_NOTIMPL;
 }
@@ -3165,8 +3228,9 @@ static HRESULT WINAPI OleObject_GetMoniker(IOleObject *iface, DWORD assign, DWOR
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%ld %ld %p)\n", This, assign, which, moniker);
+    TRACE("(%p)->(%ld %ld %p)\n", This, assign, which, moniker);
 
+    if (moniker) *moniker = NULL;
     return E_NOTIMPL;
 }
 
@@ -3175,7 +3239,7 @@ static HRESULT WINAPI OleObject_InitFromData(IOleObject *iface, IDataObject *dat
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%p %d %ld)\n", This, dataobj, creation, reserved);
+    TRACE("(%p)->(%p %d %ld)\n", This, dataobj, creation, reserved);
 
     return E_NOTIMPL;
 }
@@ -3184,8 +3248,9 @@ static HRESULT WINAPI OleObject_GetClipboardData(IOleObject *iface, DWORD reserv
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%ld %p)\n", This, reserved, dataobj);
+    TRACE("(%p)->(%ld %p)\n", This, reserved, dataobj);
 
+    if (dataobj) *dataobj = NULL;
     return E_NOTIMPL;
 }
 
@@ -3194,63 +3259,80 @@ static HRESULT WINAPI OleObject_DoVerb(IOleObject *iface, LONG verb, LPMSG msg,
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%ld %p %p %ld %p %p)\n", This, verb, msg, active_site, index, hwndParent, rect);
+    TRACE("(%p)->(%ld %p %p %ld %p %p)\n", This, verb, msg, active_site, index, hwndParent, rect);
 
-    return E_NOTIMPL;
+    switch (verb)
+    {
+    case OLEIVERB_PRIMARY:
+    case OLEIVERB_SHOW:
+    case OLEIVERB_HIDE:
+    case OLEIVERB_INPLACEACTIVATE:
+    case OLEIVERB_UIACTIVATE:
+    case OLEIVERB_DISCARDUNDOSTATE:
+        /* the control has no window: there is nothing to show or activate */
+        return S_OK;
+    default:
+        return verb > 0 ? OLEOBJ_S_INVALIDVERB : E_NOTIMPL;
+    }
 }
 
 static HRESULT WINAPI OleObject_EnumVerbs(IOleObject *iface, IEnumOLEVERB **enumoleverb)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%p)\n", This, enumoleverb);
+    TRACE("(%p)->(%p)\n", This, enumoleverb);
 
-    return E_NOTIMPL;
+    return OleRegEnumVerbs(&CLSID_ScriptControl, enumoleverb);
 }
 
 static HRESULT WINAPI OleObject_Update(IOleObject *iface)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)\n", This);
+    TRACE("(%p)\n", This);
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static HRESULT WINAPI OleObject_IsUpToDate(IOleObject *iface)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)\n", This);
+    TRACE("(%p)\n", This);
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static HRESULT WINAPI OleObject_GetUserClassID(IOleObject *iface, CLSID *clsid)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%p)\n", This, clsid);
+    TRACE("(%p)->(%p)\n", This, clsid);
 
-    return E_NOTIMPL;
+    if (!clsid) return E_POINTER;
+    *clsid = CLSID_ScriptControl;
+    return S_OK;
 }
 
 static HRESULT WINAPI OleObject_GetUserType(IOleObject *iface, DWORD form_of_type, LPOLESTR *usertype)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%ld %p)\n", This, form_of_type, usertype);
+    TRACE("(%p)->(%ld %p)\n", This, form_of_type, usertype);
 
-    return E_NOTIMPL;
+    return OleRegGetUserType(&CLSID_ScriptControl, form_of_type, usertype);
 }
 
 static HRESULT WINAPI OleObject_SetExtent(IOleObject *iface, DWORD aspect, SIZEL *size)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%ld %p)\n", This, aspect, size);
+    TRACE("(%p)->(%ld %p)\n", This, aspect, size);
 
-    return E_NOTIMPL;
+    if (aspect != DVASPECT_CONTENT) return DV_E_DVASPECT;
+    if (!size) return E_POINTER;
+    This->extent = *size;
+    return S_OK;
 }
 
 static HRESULT WINAPI OleObject_GetExtent(IOleObject *iface, DWORD aspect, SIZEL *size)
@@ -3269,28 +3351,34 @@ static HRESULT WINAPI OleObject_GetExtent(IOleObject *iface, DWORD aspect, SIZEL
 static HRESULT WINAPI OleObject_Advise(IOleObject *iface, IAdviseSink *sink, DWORD *connection)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
+    HRESULT hr;
 
-    FIXME("(%p)->(%p %p)\n", This, sink, connection);
+    TRACE("(%p)->(%p %p)\n", This, sink, connection);
 
-    return E_NOTIMPL;
+    if (!This->advise_holder && FAILED(hr = CreateOleAdviseHolder(&This->advise_holder))) return hr;
+    return IOleAdviseHolder_Advise(This->advise_holder, sink, connection);
 }
 
 static HRESULT WINAPI OleObject_Unadvise(IOleObject *iface, DWORD connection)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%ld)\n", This, connection);
+    TRACE("(%p)->(%ld)\n", This, connection);
 
-    return E_NOTIMPL;
+    if (!This->advise_holder) return OLE_E_NOCONNECTION;
+    return IOleAdviseHolder_Unadvise(This->advise_holder, connection);
 }
 
 static HRESULT WINAPI OleObject_EnumAdvise(IOleObject *iface, IEnumSTATDATA **enumadvise)
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%p)\n", This, enumadvise);
+    TRACE("(%p)->(%p)\n", This, enumadvise);
 
-    return E_NOTIMPL;
+    if (!enumadvise) return E_POINTER;
+    *enumadvise = NULL;
+    if (!This->advise_holder) return S_OK;
+    return IOleAdviseHolder_EnumAdvise(This->advise_holder, enumadvise);
 }
 
 static HRESULT WINAPI OleObject_GetMiscStatus(IOleObject *iface, DWORD aspect, DWORD *status)
@@ -3306,7 +3394,7 @@ static HRESULT WINAPI OleObject_SetColorScheme(IOleObject *iface, LOGPALETTE *lo
 {
     ScriptControl *This = impl_from_IOleObject(iface);
 
-    FIXME("(%p)->(%p)\n", This, logpal);
+    TRACE("(%p)->(%p)\n", This, logpal);
 
     return E_NOTIMPL;
 }
@@ -3338,6 +3426,16 @@ static const IOleObjectVtbl OleObjectVtbl = {
     OleObject_SetColorScheme
 };
 
+
+/* the persisted state: a version, then the properties that are set by the container */
+struct persisted
+{
+    DWORD version;
+    LONG timeout;
+    VARIANT_BOOL allow_ui;
+    VARIANT_BOOL use_safe_subset;
+};
+
 static HRESULT WINAPI PersistStreamInit_QueryInterface(IPersistStreamInit *iface, REFIID riid, void **obj)
 {
     ScriptControl *This = impl_from_IPersistStreamInit(iface);
@@ -3360,52 +3458,75 @@ static HRESULT WINAPI PersistStreamInit_GetClassID(IPersistStreamInit *iface, CL
 {
     ScriptControl *This = impl_from_IPersistStreamInit(iface);
 
-    FIXME("(%p)->(%p)\n", This, clsid);
+    TRACE("(%p)->(%p)\n", This, clsid);
 
-    return E_NOTIMPL;
+    if (!clsid) return E_POINTER;
+    *clsid = CLSID_ScriptControl;
+    return S_OK;
 }
 
 static HRESULT WINAPI PersistStreamInit_IsDirty(IPersistStreamInit *iface)
 {
     ScriptControl *This = impl_from_IPersistStreamInit(iface);
 
-    FIXME("(%p)\n", This);
+    TRACE("(%p)\n", This);
 
-    return E_NOTIMPL;
+    return This->dirty ? S_OK : S_FALSE;
 }
 
 static HRESULT WINAPI PersistStreamInit_Load(IPersistStreamInit *iface, IStream *stream)
 {
     ScriptControl *This = impl_from_IPersistStreamInit(iface);
+    struct persisted data;
+    ULONG read;
+    HRESULT hr;
 
-    FIXME("(%p)->(%p)\n", This, stream);
+    TRACE("(%p)->(%p)\n", This, stream);
 
+    if (!stream) return E_POINTER;
+    hr = IStream_Read(stream, &data, sizeof(data), &read);
+    if (FAILED(hr)) return hr;
+    if (read != sizeof(data) || data.version != 1) return E_FAIL;
+    This->timeout = data.timeout;
+    This->allow_ui = data.allow_ui;
+    This->use_safe_subset = data.use_safe_subset;
+    This->dirty = FALSE;
     return S_OK;
 }
 
 static HRESULT WINAPI PersistStreamInit_Save(IPersistStreamInit *iface, IStream *stream, BOOL clear_dirty)
 {
     ScriptControl *This = impl_from_IPersistStreamInit(iface);
+    struct persisted data;
+    HRESULT hr;
 
-    FIXME("(%p)->(%p %d)\n", This, stream, clear_dirty);
+    TRACE("(%p)->(%p %d)\n", This, stream, clear_dirty);
 
-    return E_NOTIMPL;
+    if (!stream) return E_POINTER;
+    data.version = 1;
+    data.timeout = This->timeout;
+    data.allow_ui = This->allow_ui;
+    data.use_safe_subset = This->use_safe_subset;
+    if (SUCCEEDED(hr = IStream_Write(stream, &data, sizeof(data), NULL)) && clear_dirty) This->dirty = FALSE;
+    return hr;
 }
 
 static HRESULT WINAPI PersistStreamInit_GetSizeMax(IPersistStreamInit *iface, ULARGE_INTEGER *size)
 {
     ScriptControl *This = impl_from_IPersistStreamInit(iface);
 
-    FIXME("(%p)->(%p)\n", This, size);
+    TRACE("(%p)->(%p)\n", This, size);
 
-    return E_NOTIMPL;
+    if (!size) return E_POINTER;
+    size->QuadPart = sizeof(struct persisted);
+    return S_OK;
 }
 
 static HRESULT WINAPI PersistStreamInit_InitNew(IPersistStreamInit *iface)
 {
     ScriptControl *This = impl_from_IPersistStreamInit(iface);
 
-    FIXME("(%p)\n", This);
+    TRACE("(%p)\n", This);
 
     return S_OK;
 }
@@ -3459,16 +3580,16 @@ static HRESULT WINAPI OleControl_OnMnemonic(IOleControl *iface, MSG *msg)
 {
     ScriptControl *This = impl_from_IOleControl(iface);
 
-    FIXME("(%p)->(%p)\n", This, msg);
+    TRACE("(%p)->(%p)\n", This, msg);
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static HRESULT WINAPI OleControl_OnAmbientPropertyChange(IOleControl *iface, DISPID dispid)
 {
     ScriptControl *This = impl_from_IOleControl(iface);
 
-    FIXME("(%p)->(%#lx)\n", This, dispid);
+    TRACE("(%p)->(%#lx)\n", This, dispid);
 
     return S_OK;
 }
@@ -3477,8 +3598,9 @@ static HRESULT WINAPI OleControl_FreezeEvents(IOleControl *iface, BOOL freeze)
 {
     ScriptControl *This = impl_from_IOleControl(iface);
 
-    FIXME("(%p)->(%d)\n", This, freeze);
+    TRACE("(%p)->(%d)\n", This, freeze);
 
+    This->events_frozen = freeze;
     return S_OK;
 }
 
@@ -3513,28 +3635,54 @@ static ULONG WINAPI QuickActivate_Release(IQuickActivate *iface)
 static HRESULT WINAPI QuickActivate_QuickActivate(IQuickActivate *iface, QACONTAINER *container, QACONTROL *control)
 {
     ScriptControl *This = impl_from_IQuickActivate(iface);
+    IConnectionPoint *cp;
+    HRESULT hr;
 
-    FIXME("(%p)->(%p %p)\n", This, container, control);
+    TRACE("(%p)->(%p %p)\n", This, container, control);
 
-    return E_NOTIMPL;
+    if (!container || !control) return E_POINTER;
+    if (container->cbSize < sizeof(*container) || control->cbSize < sizeof(*control)) return E_INVALIDARG;
+
+    if (container->pClientSite) IOleObject_SetClientSite(&This->IOleObject_iface, container->pClientSite);
+
+    /* the container's sinks for the control's events and property notifications */
+    if (container->pUnkEventSink)
+    {
+        cp = &This->cp_scsource.IConnectionPoint_iface;
+        if (SUCCEEDED(hr = IConnectionPoint_Advise(cp, container->pUnkEventSink, &control->dwEventCookie)))
+            TRACE("event sink cookie %lu\n", control->dwEventCookie);
+    }
+    if (container->pPropertyNotifySink)
+    {
+        cp = &This->cp_propnotif.IConnectionPoint_iface;
+        IConnectionPoint_Advise(cp, (IUnknown *)container->pPropertyNotifySink, &control->dwPropNotifyCookie);
+    }
+    if (container->pAdviseSink)
+        { DWORD adv; IOleObject_Advise(&This->IOleObject_iface, (IAdviseSink *)container->pAdviseSink, &adv); }
+
+    control->dwMiscStatus = 0;
+    OleRegGetMiscStatus(&CLSID_ScriptControl, DVASPECT_CONTENT, &control->dwMiscStatus);
+    control->dwViewStatus = VIEWSTATUS_OPAQUE;
+    control->dwPointerActivationPolicy = 0;
+    return S_OK;
 }
 
 static HRESULT WINAPI QuickActivate_SetContentExtent(IQuickActivate *iface, SIZEL *size)
 {
     ScriptControl *This = impl_from_IQuickActivate(iface);
 
-    FIXME("(%p)->(%p)\n", This, size);
+    TRACE("(%p)->(%p)\n", This, size);
 
-    return E_NOTIMPL;
+    return IOleObject_SetExtent(&This->IOleObject_iface, DVASPECT_CONTENT, size);
 }
 
 static HRESULT WINAPI QuickActivate_GetContentExtent(IQuickActivate *iface, SIZEL *size)
 {
     ScriptControl *This = impl_from_IQuickActivate(iface);
 
-    FIXME("(%p)->(%p)\n", This, size);
+    TRACE("(%p)->(%p)\n", This, size);
 
-    return E_NOTIMPL;
+    return IOleObject_GetExtent(&This->IOleObject_iface, DVASPECT_CONTENT, size);
 }
 
 static const IQuickActivateVtbl QuickActivateVtbl = {
@@ -3570,10 +3718,12 @@ static HRESULT WINAPI ViewObject_Draw(IViewObjectEx *iface, DWORD drawaspect, LO
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
 
-    FIXME("(%p)->(%ld %ld %p %p %p %p %p %p %p %Iu)\n", This, drawaspect, index, aspect, device, target_dev,
+    TRACE("(%p)->(%ld %ld %p %p %p %p %p %p %p %Iu)\n", This, drawaspect, index, aspect, device, target_dev,
         hdc_draw, bounds, win_bounds, fn_continue, cont);
 
-    return E_NOTIMPL;
+    if (drawaspect != DVASPECT_CONTENT) return DV_E_DVASPECT;
+    /* a control that is invisible at run time draws nothing */
+    return S_OK;
 }
 
 static HRESULT WINAPI ViewObject_GetColorSet(IViewObjectEx *iface, DWORD drawaspect, LONG index, void *aspect,
@@ -3581,10 +3731,10 @@ static HRESULT WINAPI ViewObject_GetColorSet(IViewObjectEx *iface, DWORD drawasp
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
 
-    FIXME("(%p)->(%ld %ld %p %p %p %p)\n", This, drawaspect, index, aspect, device, hic_target,
-        colorset);
+    TRACE("(%p)->(%ld %ld %p %p %p %p)\n", This, drawaspect, index, aspect, device, hic_target, colorset);
 
-    return E_NOTIMPL;
+    if (colorset) *colorset = NULL;
+    return S_FALSE;
 }
 
 static HRESULT WINAPI ViewObject_Freeze(IViewObjectEx *iface, DWORD drawaspect, LONG index, void *aspect,
@@ -3592,18 +3742,20 @@ static HRESULT WINAPI ViewObject_Freeze(IViewObjectEx *iface, DWORD drawaspect,
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
 
-    FIXME("(%p)->(%ld %ld %p %p)\n", This, drawaspect, index, aspect, freeze);
+    TRACE("(%p)->(%ld %ld %p %p)\n", This, drawaspect, index, aspect, freeze);
 
-    return E_NOTIMPL;
+    if (drawaspect != DVASPECT_CONTENT) return DV_E_DVASPECT;
+    if (freeze) *freeze = 1;
+    return S_OK;
 }
 
 static HRESULT WINAPI ViewObject_Unfreeze(IViewObjectEx *iface, DWORD freeze)
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
 
-    FIXME("(%p)->(%ld)\n", This, freeze);
+    TRACE("(%p)->(%ld)\n", This, freeze);
 
-    return E_NOTIMPL;
+    return freeze == 1 ? S_OK : OLE_E_NOCONNECTION;
 }
 
 static HRESULT WINAPI ViewObject_SetAdvise(IViewObjectEx *iface, DWORD aspects, DWORD flags, IAdviseSink *sink)
@@ -3649,18 +3801,26 @@ static HRESULT WINAPI ViewObject_GetExtent(IViewObjectEx *iface, DWORD draw_aspe
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
 
-    FIXME("(%p)->(%ld %ld %p %p)\n", This, draw_aspect, index, device, size);
+    TRACE("(%p)->(%ld %ld %p %p)\n", This, draw_aspect, index, device, size);
 
-    return E_NOTIMPL;
+    if (draw_aspect != DVASPECT_CONTENT) return DV_E_DVASPECT;
+    if (!size) return E_POINTER;
+    *size = This->extent;
+    return S_OK;
 }
 
 static HRESULT WINAPI ViewObject_GetRect(IViewObjectEx *iface, DWORD aspect, RECTL *rect)
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
 
-    FIXME("(%p)->(%ld %p)\n", This, aspect, rect);
+    TRACE("(%p)->(%ld %p)\n", This, aspect, rect);
 
-    return E_NOTIMPL;
+    if (aspect != DVASPECT_CONTENT) return DV_E_DVASPECT;
+    if (!rect) return E_POINTER;
+    rect->left = rect->top = 0;
+    rect->right = This->extent.cx;
+    rect->bottom = This->extent.cy;
+    return S_OK;
 }
 
 static HRESULT WINAPI ViewObject_GetViewStatus(IViewObjectEx *iface, DWORD *status)
@@ -3678,19 +3838,26 @@ static HRESULT WINAPI ViewObject_QueryHitPoint(IViewObjectEx *iface, DWORD aspec
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
 
-    FIXME("(%p)->(%ld %s %s %ld %p)\n", This, aspect, wine_dbgstr_rect(bounds), wine_dbgstr_point(&pt), close_hint, hit_result);
+    TRACE("(%p)->(%ld %s %s %ld %p)\n", This, aspect, wine_dbgstr_rect(bounds), wine_dbgstr_point(&pt), close_hint, hit_result);
 
-    return E_NOTIMPL;
+    if (aspect != DVASPECT_CONTENT) return DV_E_DVASPECT;
+    if (!bounds || !hit_result) return E_POINTER;
+    *hit_result = PtInRect(bounds, pt) ? HITRESULT_HIT : HITRESULT_OUTSIDE;
+    return S_OK;
 }
 
 static HRESULT WINAPI ViewObject_QueryHitRect(IViewObjectEx *iface, DWORD aspect, const RECT *bounds,
     const RECT *loc, LONG close_hint, DWORD *hit_result)
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
+    RECT inter;
 
-    FIXME("(%p)->(%ld %s %s %ld %p)\n", This, aspect, wine_dbgstr_rect(bounds), wine_dbgstr_rect(loc), close_hint, hit_result);
+    TRACE("(%p)->(%ld %s %s %ld %p)\n", This, aspect, wine_dbgstr_rect(bounds), wine_dbgstr_rect(loc), close_hint, hit_result);
 
-    return E_NOTIMPL;
+    if (aspect != DVASPECT_CONTENT) return DV_E_DVASPECT;
+    if (!bounds || !loc || !hit_result) return E_POINTER;
+    *hit_result = IntersectRect(&inter, bounds, loc) ? HITRESULT_HIT : HITRESULT_OUTSIDE;
+    return S_OK;
 }
 
 static HRESULT WINAPI ViewObject_GetNaturalExtent(IViewObjectEx *iface, DWORD aspect, LONG index,
@@ -3698,9 +3865,12 @@ static HRESULT WINAPI ViewObject_GetNaturalExtent(IViewObjectEx *iface, DWORD as
 {
     ScriptControl *This = impl_from_IViewObjectEx(iface);
 
-    FIXME("(%p)->(%ld %ld %p %p %p %p)\n", This, aspect, index, device, target_hdc, extent_info, size);
+    TRACE("(%p)->(%ld %ld %p %p %p %p)\n", This, aspect, index, device, target_hdc, extent_info, size);
 
-    return E_NOTIMPL;
+    if (aspect != DVASPECT_CONTENT) return DV_E_DVASPECT;
+    if (!size) return E_POINTER;
+    *size = This->extent;
+    return S_OK;
 }
 
 static const IViewObjectExVtbl ViewObjectExVtbl = {
@@ -3757,9 +3927,9 @@ static HRESULT WINAPI PointerInactive_OnInactiveMouseMove(IPointerInactive *ifac
 {
     ScriptControl *This = impl_from_IPointerInactive(iface);
 
-    FIXME("(%p)->(%s %ld %ld %#lx)\n", This, wine_dbgstr_rect(bounds), x, y, key_state);
+    TRACE("(%p)->(%s %ld %ld %#lx)\n", This, wine_dbgstr_rect(bounds), x, y, key_state);
 
-    return E_NOTIMPL;
+    return S_OK;
 }
 
 static HRESULT WINAPI PointerInactive_OnInactiveSetCursor(IPointerInactive *iface, const RECT *bounds,
@@ -3767,9 +3937,10 @@ static HRESULT WINAPI PointerInactive_OnInactiveSetCursor(IPointerInactive *ifac
 {
     ScriptControl *This = impl_from_IPointerInactive(iface);
 
-    FIXME("(%p)->(%s %ld %ld %#lx %d)\n", This, wine_dbgstr_rect(bounds), x, y, msg, set_always);
+    TRACE("(%p)->(%s %ld %ld %#lx %d)\n", This, wine_dbgstr_rect(bounds), x, y, msg, set_always);
 
-    return E_NOTIMPL;
+    /* the cursor is the container's */
+    return S_FALSE;
 }
 
 static const IPointerInactiveVtbl PointerInactiveVtbl = {
@@ -3781,6 +3952,240 @@ static const IPointerInactiveVtbl PointerInactiveVtbl = {
     PointerInactive_OnInactiveSetCursor
 };
 
+
+/* IEnumConnections over a copy of the connections of a connection point */
+struct enum_connections
+{
+    IEnumConnections IEnumConnections_iface;
+    LONG ref;
+    CONNECTDATA *data;
+    ULONG count, pos;
+};
+
+static inline struct enum_connections *impl_from_IEnumConnections(IEnumConnections *iface)
+{
+    return CONTAINING_RECORD(iface, struct enum_connections, IEnumConnections_iface);
+}
+
+static HRESULT WINAPI enum_connections_QueryInterface(IEnumConnections *iface, REFIID riid, void **obj)
+{
+    if (IsEqualGUID(riid, &IID_IUnknown) || IsEqualGUID(riid, &IID_IEnumConnections))
+    {
+        IEnumConnections_AddRef((*obj = iface));
+        return S_OK;
+    }
+    *obj = NULL;
+    return E_NOINTERFACE;
+}
+
+static ULONG WINAPI enum_connections_AddRef(IEnumConnections *iface)
+{
+    return InterlockedIncrement(&impl_from_IEnumConnections(iface)->ref);
+}
+
+static ULONG WINAPI enum_connections_Release(IEnumConnections *iface)
+{
+    struct enum_connections *This = impl_from_IEnumConnections(iface);
+    LONG ref = InterlockedDecrement(&This->ref);
+    ULONG i;
+
+    if (!ref)
+    {
+        for (i = 0; i < This->count; i++) IUnknown_Release(This->data[i].pUnk);
+        free(This->data);
+        free(This);
+    }
+    return ref;
+}
+
+static HRESULT WINAPI enum_connections_Next(IEnumConnections *iface, ULONG count, CONNECTDATA *ret, ULONG *fetched)
+{
+    struct enum_connections *This = impl_from_IEnumConnections(iface);
+    ULONG n = 0;
+
+    if (!ret || (!fetched && count != 1)) return E_POINTER;
+    while (n < count && This->pos < This->count)
+    {
+        ret[n] = This->data[This->pos++];
+        IUnknown_AddRef(ret[n].pUnk);
+        n++;
+    }
+    if (fetched) *fetched = n;
+    return n == count ? S_OK : S_FALSE;
+}
+
+static HRESULT WINAPI enum_connections_Skip(IEnumConnections *iface, ULONG count)
+{
+    struct enum_connections *This = impl_from_IEnumConnections(iface);
+
+    if (count > This->count - This->pos)
+    {
+        This->pos = This->count;
+        return S_FALSE;
+    }
+    This->pos += count;
+    return S_OK;
+}
+
+static HRESULT WINAPI enum_connections_Reset(IEnumConnections *iface)
+{
+    impl_from_IEnumConnections(iface)->pos = 0;
+    return S_OK;
+}
+
+static const IEnumConnectionsVtbl enum_connections_vtbl;
+
+static HRESULT enum_connections_create(const CONNECTDATA *data, ULONG count, ULONG pos, IEnumConnections **ret)
+{
+    struct enum_connections *e;
+    ULONG i;
+
+    if (!(e = calloc(1, sizeof(*e)))) return E_OUTOFMEMORY;
+    if (count && !(e->data = calloc(count, sizeof(*e->data))))
+    {
+        free(e);
+        return E_OUTOFMEMORY;
+    }
+    e->IEnumConnections_iface.lpVtbl = &enum_connections_vtbl;
+    e->ref = 1;
+    e->count = count;
+    e->pos = pos;
+    for (i = 0; i < count; i++)
+    {
+        e->data[i] = data[i];
+        IUnknown_AddRef(e->data[i].pUnk);
+    }
+    *ret = &e->IEnumConnections_iface;
+    return S_OK;
+}
+
+static HRESULT WINAPI enum_connections_Clone(IEnumConnections *iface, IEnumConnections **ret)
+{
+    struct enum_connections *This = impl_from_IEnumConnections(iface);
+
+    if (!ret) return E_POINTER;
+    return enum_connections_create(This->data, This->count, This->pos, ret);
+}
+
+static const IEnumConnectionsVtbl enum_connections_vtbl =
+{
+    enum_connections_QueryInterface, enum_connections_AddRef, enum_connections_Release,
+    enum_connections_Next, enum_connections_Skip, enum_connections_Reset, enum_connections_Clone,
+};
+
+/* IEnumConnectionPoints over the connection points of the control */
+struct enum_points
+{
+    IEnumConnectionPoints IEnumConnectionPoints_iface;
+    LONG ref;
+    IConnectionPoint *points[2];
+    ULONG count, pos;
+};
+
+static inline struct enum_points *impl_from_IEnumConnectionPoints(IEnumConnectionPoints *iface)
+{
+    return CONTAINING_RECORD(iface, struct enum_points, IEnumConnectionPoints_iface);
+}
+
+static HRESULT WINAPI enum_points_QueryInterface(IEnumConnectionPoints *iface, REFIID riid, void **obj)
+{
+    if (IsEqualGUID(riid, &IID_IUnknown) || IsEqualGUID(riid, &IID_IEnumConnectionPoints))
+    {
+        IEnumConnectionPoints_AddRef((*obj = iface));
+        return S_OK;
+    }
+    *obj = NULL;
+    return E_NOINTERFACE;
+}
+
+static ULONG WINAPI enum_points_AddRef(IEnumConnectionPoints *iface)
+{
+    return InterlockedIncrement(&impl_from_IEnumConnectionPoints(iface)->ref);
+}
+
+static ULONG WINAPI enum_points_Release(IEnumConnectionPoints *iface)
+{
+    struct enum_points *This = impl_from_IEnumConnectionPoints(iface);
+    LONG ref = InterlockedDecrement(&This->ref);
+    ULONG i;
+
+    if (!ref)
+    {
+        for (i = 0; i < This->count; i++) IConnectionPoint_Release(This->points[i]);
+        free(This);
+    }
+    return ref;
+}
+
+static HRESULT WINAPI enum_points_Next(IEnumConnectionPoints *iface, ULONG count, IConnectionPoint **ret, ULONG *fetched)
+{
+    struct enum_points *This = impl_from_IEnumConnectionPoints(iface);
+    ULONG n = 0;
+
+    if (!ret || (!fetched && count != 1)) return E_POINTER;
+    while (n < count && This->pos < This->count)
+    {
+        IConnectionPoint_AddRef((ret[n] = This->points[This->pos++]));
+        n++;
+    }
+    if (fetched) *fetched = n;
+    return n == count ? S_OK : S_FALSE;
+}
+
+static HRESULT WINAPI enum_points_Skip(IEnumConnectionPoints *iface, ULONG count)
+{
+    struct enum_points *This = impl_from_IEnumConnectionPoints(iface);
+
+    if (count > This->count - This->pos)
+    {
+        This->pos = This->count;
+        return S_FALSE;
+    }
+    This->pos += count;
+    return S_OK;
+}
+
+static HRESULT WINAPI enum_points_Reset(IEnumConnectionPoints *iface)
+{
+    impl_from_IEnumConnectionPoints(iface)->pos = 0;
+    return S_OK;
+}
+
+static const IEnumConnectionPointsVtbl enum_points_vtbl;
+
+static HRESULT enum_points_create(IConnectionPoint **points, ULONG count, ULONG pos, IEnumConnectionPoints **ret)
+{
+    struct enum_points *e;
+    ULONG i;
+
+    if (!(e = calloc(1, sizeof(*e)))) return E_OUTOFMEMORY;
+    e->IEnumConnectionPoints_iface.lpVtbl = &enum_points_vtbl;
+    e->ref = 1;
+    e->count = count;
+    e->pos = pos;
+    for (i = 0; i < count; i++)
+    {
+        e->points[i] = points[i];
+        IConnectionPoint_AddRef(points[i]);
+    }
+    *ret = &e->IEnumConnectionPoints_iface;
+    return S_OK;
+}
+
+static HRESULT WINAPI enum_points_Clone(IEnumConnectionPoints *iface, IEnumConnectionPoints **ret)
+{
+    struct enum_points *This = impl_from_IEnumConnectionPoints(iface);
+
+    if (!ret) return E_POINTER;
+    return enum_points_create(This->points, This->count, This->pos, ret);
+}
+
+static const IEnumConnectionPointsVtbl enum_points_vtbl =
+{
+    enum_points_QueryInterface, enum_points_AddRef, enum_points_Release,
+    enum_points_Next, enum_points_Skip, enum_points_Reset, enum_points_Clone,
+};
+
 static HRESULT WINAPI ConnectionPointContainer_QueryInterface(IConnectionPointContainer *iface, REFIID riid, void **obj)
 {
     ScriptControl *This = impl_from_IConnectionPointContainer(iface);
@@ -3802,10 +4207,13 @@ static ULONG WINAPI ConnectionPointContainer_Release(IConnectionPointContainer *
 static HRESULT WINAPI ConnectionPointContainer_EnumConnectionPoints(IConnectionPointContainer *iface, IEnumConnectionPoints **enum_points)
 {
     ScriptControl *This = impl_from_IConnectionPointContainer(iface);
+    IConnectionPoint *points[2] = { &This->cp_scsource.IConnectionPoint_iface,
+                                    &This->cp_propnotif.IConnectionPoint_iface };
 
-    FIXME("(%p)->(%p)\n", This, enum_points);
+    TRACE("(%p)->(%p)\n", This, enum_points);
 
-    return E_NOTIMPL;
+    if (!enum_points) return E_POINTER;
+    return enum_points_create(points, 2, 0, enum_points);
 }
 
 static HRESULT WINAPI ConnectionPointContainer_FindConnectionPoint(IConnectionPointContainer *iface, REFIID riid, IConnectionPoint **cp)
@@ -3902,29 +4310,76 @@ static HRESULT WINAPI ConnectionPoint_Advise(IConnectionPoint *iface, IUnknown *
         DWORD *cookie)
 {
     ConnectionPoint *This = impl_from_IConnectionPoint(iface);
+    IUnknown *sink;
+    struct sink *sinks;
+    HRESULT hr;
 
-    FIXME("(%p)->(%p %p)\n", This, unk_sink, cookie);
+    TRACE("(%p)->(%p %p)\n", This, unk_sink, cookie);
 
-    return E_NOTIMPL;
+    if (!unk_sink || !cookie) return E_POINTER;
+    *cookie = 0;
+
+    /* the events are a dispinterface, which the sink may implement as its IDispatch */
+    if (FAILED(hr = IUnknown_QueryInterface(unk_sink, This->riid, (void **)&sink)))
+    {
+        if (!IsEqualGUID(This->riid, &IID_IPropertyNotifySink) &&
+            SUCCEEDED(IUnknown_QueryInterface(unk_sink, &IID_IDispatch, (void **)&sink)))
+            hr = S_OK;
+        else
+            return CONNECT_E_CANNOTCONNECT;
+    }
+    if (!(sinks = realloc(This->sinks, (This->sink_count + 1) * sizeof(*sinks))))
+    {
+        IUnknown_Release(sink);
+        return E_OUTOFMEMORY;
+    }
+    This->sinks = sinks;
+    if (!This->next_cookie) This->next_cookie = 1;
+    This->sinks[This->sink_count].cookie = This->next_cookie;
+    This->sinks[This->sink_count].unk = sink;
+    This->sink_count++;
+    *cookie = This->next_cookie++;
+    return S_OK;
 }
 
 static HRESULT WINAPI ConnectionPoint_Unadvise(IConnectionPoint *iface, DWORD cookie)
 {
     ConnectionPoint *This = impl_from_IConnectionPoint(iface);
+    unsigned int i;
 
-    FIXME("(%p)->(%ld)\n", This, cookie);
+    TRACE("(%p)->(%ld)\n", This, cookie);
 
-    return E_NOTIMPL;
+    for (i = 0; i < This->sink_count; i++)
+    {
+        if (This->sinks[i].cookie != cookie) continue;
+        IUnknown_Release(This->sinks[i].unk);
+        memmove(&This->sinks[i], &This->sinks[i + 1], (This->sink_count - i - 1) * sizeof(*This->sinks));
+        This->sink_count--;
+        return S_OK;
+    }
+    return CONNECT_E_NOCONNECTION;
 }
 
 static HRESULT WINAPI ConnectionPoint_EnumConnections(IConnectionPoint *iface,
         IEnumConnections **ppEnum)
 {
     ConnectionPoint *This = impl_from_IConnectionPoint(iface);
+    CONNECTDATA *data = NULL;
+    unsigned int i;
+    HRESULT hr;
 
-    FIXME("(%p)->(%p): stub\n", This, ppEnum);
+    TRACE("(%p)->(%p)\n", This, ppEnum);
 
-    return E_NOTIMPL;
+    if (!ppEnum) return E_POINTER;
+    if (This->sink_count && !(data = calloc(This->sink_count, sizeof(*data)))) return E_OUTOFMEMORY;
+    for (i = 0; i < This->sink_count; i++)
+    {
+        data[i].pUnk = This->sinks[i].unk;
+        data[i].dwCookie = This->sinks[i].cookie;
+    }
+    hr = enum_connections_create(data, This->sink_count, 0, ppEnum);
+    free(data);
+    return hr;
 }
 
 static const IConnectionPointVtbl ConnectionPointVtbl =
